Kwyss/new shape owns data

kwyss-nvidia opened the pull request:
# Description

Reapply changes to have nvte_shape own its data.
This time, replace the pointer with an array. The previous change opted for backward compatibility, but there were cases where the default copy constructor preserved indirection to expiring data.

This change makes the data array the only source of truth. Array to pointer decay allows code changes to be relatively minimal to update to the new struct.


## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

nvte_shape owns its shape data.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

kwyss-nvidia commented:
/te-ci

timmoon10 reviewed (commented):
Thanks for the fix! I am seeing some C++ test failures because the compiler can't disambiguate between `NVTEShape` and `std::vector<size_t>`. The other test failures seem orthogonal to this PR (the nightly build is also seeing errors in `test_fused_attn.py`, other PRs are also seeing JAX build failures).

For future reference, an initial attempt at redesigning `NVTEShape` was merged in https://github.com/NVIDIA/TransformerEngine/pull/1674 and reverted with https://github.com/NVIDIA/TransformerEngine/pull/1703.

kwyss-nvidia commented:
> Thanks for the fix! I am seeing some C++ test failures because the compiler can't disambiguate between `NVTEShape` and `std::vector<size_t>`. The other test failures seem orthogonal to this PR (the nightly build is also seeing errors in `test_fused_attn.py`, other PRs are also seeing JAX build failures).
> 
> For future reference, an initial attempt at redesigning `NVTEShape` was merged in #1674 and reverted with #1703.

Thanks Tim. I added the qualifiers to build the CPP tests locally with `qa/L0_cppunittest/test.sh`. Forgot that that compile is separate from the main install.

kwyss-nvidia commented:
/te-ci
